OWML.ModHelper/Logging/ModSocketOutput.cs

```diff
@@ -1,56 +1,97 @@
 ﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
 using System.Linq;
 using System.Net;
 using System.Net.Sockets;
 using System.Text;
+using Newtonsoft.Json;
 using OWML.Common;
 
 namespace OWML.ModHelper
 {
     public class ModSocketOutput : ModConsole
     {
-        private const string LocalHost = "127.0.0.1";
-        private const string NameMessageSeparator = ";;";
-
+        private readonly int _port;
         private static Socket _socket;
+        private static IModLogger _logger;
 
         public ModSocketOutput(IOwmlConfig config, IModLogger logger, IModManifest manifest) : base(config, logger, manifest)
         {
             if (_socket == null)
             {
-                CreateSocket();
+                _port = config.SocketPort;
+                ConnectToSocket();
+                _logger = logger;
             }
         }
 
-        public override void WriteLine(string s)
+        [Obsolete("Use WriteLine(string) or WriteLine(string, MessageType) instead.")]
+        public override void WriteLine(params object[] objects)
         {
-            Logger?.Log(s);
-            CallWriteCallback(Manifest, s);
-            var message = $"{Manifest.Name}{NameMessageSeparator}{s}";
-            InternalWriteLine(message);
+            var line = string.Join(" ", objects.Select(o => o.ToString()).ToArray());
+            var type = MessageType.Message;
+            WriteLine(type, line, GetCallingMethodName(new StackTrace()));
         }
 
-        public override void WriteLine(params object[] objects)
+        public override void WriteLine(string line)
         {
-            WriteLine(string.Join(" ", objects.Select(o => o.ToString()).ToArray()));
+            var type = MessageType.Message;
+            WriteLine(type, line, GetCallingMethodName(new StackTrace()));
         }
 
-        private void CreateSocket()
+        public override void WriteLine(string line, MessageType type)
         {
-            var consolePortArgument = CommandLineArguments.GetArgument(Constants.ConsolePortArgument);
-            if (!int.TryParse(consolePortArgument, out var port))
+            WriteLine(type, line, GetCallingMethodName(new StackTrace()));
+        }
+
+        private void WriteLine(MessageType type, string line, string senderType)
+        {
+            Logger?.Log(line);
+            CallWriteCallback(Manifest, line);
+
+            var message = new SocketMessage
             {
-                Logger?.Log("Error: Missing or incorrectly formatted console port argument");
-                return;
+                SenderName = Manifest.Name,
+                SenderType = senderType,
+                Type = type,
+                Message = line
+            };
+            var json = JsonConvert.SerializeObject(message);
+
+            WriteToSocket(json);
+        }
+
+        private string GetCallingMethodName(StackTrace frame)
```

Comment:
"GetCallingMethodName" makes it sound like it's getting the name of a method, but it's actually getting the name of the class that defines the method, right?

Reply:
renamed to getcallingtype